src/netbase.cpp

```diff
@@ -1236,55 +1236,85 @@ CSubNet::CSubNet() : valid(false)
     memset(netmask, 0, sizeof(netmask));
 }
 
-CSubNet::CSubNet(const std::string& strSubnet)
+CSubNet::CSubNet(const CNetAddr& addr, int32_t mask)
 {
-    size_t slash = strSubnet.find_last_of('/');
-    std::vector<CNetAddr> vIP;
+    valid = true;
+    network = addr;
+    // Default to /32 (IPv4) or /128 (IPv6), i.e. match single address
+    memset(netmask, 255, sizeof(netmask));
 
+    // IPv4 addresses start at offset 12, and first 12 bytes must match, so just offset n
+    const int astartofs = network.IsIPv4() ? 12 : 0;
+
+    int32_t n = mask;
+    if(n >= 0 && n <= (128 - astartofs*8)) // Only valid if in range of bits of address
+    {
+        n += astartofs*8;
+        // Clear bits [n..127]
+        for (; n < 128; ++n)
+            netmask[n>>3] &= ~(1<<(7-(n&7)));
+    } else
+        valid = false;
+
+    // Normalize network according to netmask
+    for(int x=0; x<16; ++x)
+        network.ip[x] &= netmask[x];
+}
+
+CSubNet::CSubNet(const CNetAddr &addr, const CNetAddr &mask)
+{
     valid = true;
+    network = addr;
     // Default to /32 (IPv4) or /128 (IPv6), i.e. match single address
     memset(netmask, 255, sizeof(netmask));
 
+    // IPv4 addresses start at offset 12, and first 12 bytes must match, so just offset n
+    const int astartofs = network.IsIPv4() ? 12 : 0;
+
+    for(int x=astartofs; x<16; ++x)
+        netmask[x] = mask.ip[x];
+
+    // Normalize network according to netmask
+    for (int x = 0; x < 16; ++x)
+        network.ip[x] &= netmask[x];
+}
+
+
+bool LookupSubNet(const char* pszName, CSubNet& ret)
+{
+    std::string strSubnet(pszName);
+    size_t slash = strSubnet.find_last_of('/');
+    std::vector<CNetAddr> vIP;
+
     std::string strAddress = strSubnet.substr(0, slash);
     if (LookupHost(strAddress.c_str(), vIP, 1, false))
     {
-        network = vIP[0];
-        if (slash != strSubnet.npos) {
+        CNetAddr network = vIP[0];
+        if (slash != strSubnet.npos)
+        {
```

Comment:
styling nit: no need for this change- we usually put branches on the same line with `if`s